src/dotty/tools/dotc/typer/Namer.scala

```diff
@@ -517,18 +517,96 @@ class Namer { typer: Typer =>
       }
     }
 
-    stats foreach expand
+    stats.foreach(expand)
     mergeCompanionDefs()
     val ctxWithStats = (ctx /: stats) ((ctx, stat) => indexExpanded(stat)(ctx))
     createCompanionLinks(ctxWithStats)
     ctxWithStats
   }
 
+  /** Add all annotations of definitions in `stats` to the defined symbols */
+  def annotate(stats: List[Tree])(implicit ctx: Context): Unit = {
+    def recur(stat: Tree): Unit = stat match {
+      case pcl: PackageDef =>
+        annotate(pcl.stats)
+      case stat: untpd.MemberDef =>
+        stat.getAttachment(SymOfTree) match {
+          case Some(sym) =>
+            sym.infoOrCompleter match {
+              case info: Completer if !defn.isPredefClass(sym.owner) =>
+                // Annotate Predef methods only when they are completed;
+                // This is necessary to break a cyclic dependence between `Predef`
+                // and `deprecated` in test `compileStdLib`.
+                addAnnotations(sym, stat)(info.creationContext)
+              case _ =>
+                // Annotations were already added as part of the symbol's completion
+            }
+          case none =>
+            assert(stat.typeOpt.exists, i"no symbol for $stat")
+        }
+      case stat: untpd.Thicket =>
+        stat.trees.foreach(recur)
+      case _ =>
+    }
+
+    for (stat <- stats) recur(expanded(stat))
+  }
+
+  /** Add annotations of `stat` to `sym`.
+   *  This method can be called twice on a symbol (e.g. once
+   *  during the `annotate` phase and then again during completion).
+   *  Therefore, care needs to be taken not to add annotations again
+   *  that are already added to the symbol.
+   */
+  def addAnnotations(sym: Symbol, stat: MemberDef)(implicit ctx: Context) = {
+    // (1) The context in which an annotation of a top-evel class or module is evaluated
+    // is the closest enclosing context which has the enclosing package as owner.
+    // (2) The context in which an annotation for any other symbol is evaluated is the
+    // closest enclosing context which has the owner of the class enclpsing the symbol as owner.
+    // E.g in
+    //
+    //     package p
+    //     import a.b
+    //     class C {
+    //       import d.e
+    //       @ann m() ...
+    //     }
+    //
+    // `@ann` is evaluated in the context just outside `C`, where the `a.b`
+    // import is visible but the `d.e` import is forgotten. This measure is necessary
+    // in order to avoid cycles.
```

Comment:
@xeno-by I had lots of failing test cases in the regression tests. They failed with CyclicReference errors. I did not keep track of which ones.